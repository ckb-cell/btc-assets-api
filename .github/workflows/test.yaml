---
name: Unit Tests

on:
  pull_request:
    branches:
      - main
      - develop

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      redis:
        image: redis
        options: >-
          --health-cmd "redis-cli ping"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 6379:6379

    steps:
      - name: Checkout Code
        uses: actions/checkout@v3

      - name: Install PNPM
        uses: pnpm/action-setup@v3
        with:
          version: 8

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "21"
          cache: "pnpm"

      - name: Install dependencies
        run: pnpm install

      - name: Run Unit Tests
        env:
          JWT_SECRET: ${{ secrets.JWT_SECRET }}
          BITCOIN_JSON_RPC_URL: ${{ secrets.BITCOIN_JSON_RPC_URL }}
          BITCOIN_JSON_RPC_USERNAME: ${{ secrets.BITCOIN_JSON_RPC_USERNAME }}
          BITCOIN_JSON_RPC_PASSWORD: ${{ secrets.BITCOIN_JSON_RPC_PASSWORD }}
          BITCOIN_ELECTRS_API_URL: ${{ secrets.BITCOIN_ELECTRS_API_URL }}
          BITCOIN_SPV_SERVICE_URL: ${{ secrets.BITCOIN_SPV_SERVICE_URL }}
          CKB_RPC_URL: ${{ secrets.CKB_RPC_URL }}
          CKB_INDEXER_URL: ${{ secrets.CKB_INDEXER_URL }}
          PAYMASTER_PRIVATE_KEY: ${{ secrets.PAYMASTER_PRIVATE_KEY }}
          REDIS_URL: redis://localhost:6379
          CI_REDIS_URL: redis://localhost:6379
        run: pnpm test
