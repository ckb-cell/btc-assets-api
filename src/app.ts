import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
