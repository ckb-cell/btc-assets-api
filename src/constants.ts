export const CUSTOM_HEADERS = {
  ApiCache: 'x-api-cache',
  ResponseCacheable: 'x-response-cacheable',
};

export enum ApiCacheStatus {
  Hit = 'HIT',
  Miss = 'MISS',
}

export const JWT_IGNORE_URLS = ['/token', '/docs'];
