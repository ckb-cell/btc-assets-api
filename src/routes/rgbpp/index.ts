import { FastifyPluginCallback } from 'fastify';
import { Server } from 'http';
import container from '../../container';
import transactionRoutes from './transaction';
import { ZodTypeProvider } from 'fastify-type-provider-zod';

const rgbppRoutes: FastifyPluginCallback<Record<never, never>, Server, ZodTypeProvider> = (fastify, _, done) => {
  fastify.decorate('transactionManager', container.resolve('transactionManager'));
  fastify.register(transactionRoutes, { prefix: '/transaction' });
  done();
};

export default rgbppRoutes;
