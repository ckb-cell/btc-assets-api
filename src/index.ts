import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
