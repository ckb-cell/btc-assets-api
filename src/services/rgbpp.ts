import { UTXO } from './bitcoin/schema';
import pLimit from 'p-limit';
import asyncRetry from 'async-retry';
import { Cradle } from '../container';
import { CKBIndexerQueryOptions } from '@ckb-lumos/ckb-indexer/lib/type';
import { buildRgbppLockArgs, genRgbppLockScript } from '@rgbpp-sdk/ckb';
import * as Sentry from '@sentry/node';
import { Script } from '@ckb-lumos/lumos';
import { Job } from 'bullmq';
import { z } from 'zod';
import { Cell } from '../routes/rgbpp/types';
import BaseQueueWorker from './base/queue-worker';
import DataCache from './base/data-cache';

type RgbppUtxoCellsPair = {
  utxo: UTXO;
  cells: Cell[];
};

interface IRgbppCollectRequest {
  btcAddress: string;
  utxos: UTXO[];
}

interface IRgbppCollectJobReturn {
  [key: string]: Cell[];
}

export interface IProcessCallbacks {
  onActive?: (job: Job<IRgbppCollectRequest>) => void;
  onCompleted?: (job: Job<IRgbppCollectRequest>) => void;
  onFailed?: (job: Job<IRgbppCollectRequest> | undefined, err: Error) => void;
}

export const RGBPP_COLLECTOR_QUEUE_NAME = 'rgbpp-collector-queue';

export default class RgbppCollector extends BaseQueueWorker<IRgbppCollectRequest, IRgbppCollectJobReturn> {
  private limit: pLimit.Limit;
  private dataCahe: DataCache<IRgbppCollectJobReturn>;

  constructor(private cradle: Cradle) {
    super({
      name: RGBPP_COLLECTOR_QUEUE_NAME,
      connection: cradle.redis,
      worker: {
        lockDuration: 60_000,
        removeOnComplete: { count: 0 },
        removeOnFail: { count: 0 },
      },
    });
    this.dataCahe = new DataCache(cradle.redis, {
      prefix: 'rgbpp-collector-data',
      schema: z.record(z.array(Cell)),
    });
    this.limit = pLimit(cradle.env.CKB_RPC_MAX_CONCURRENCY);
  }

  /**
   * Get the rgbpp cells from cache
   * @param btcAddress - the btc address
   */
  public async getRgbppCellsFromCache(btcAddress: string) {
    const data = await this.dataCahe.get(btcAddress);
    if (!data) {
      return null;
    }
    return Object.values(data).flat();
  }

  /**
   * Get the cells for the utxo, use ckb indexer to query the cells
   * @param utxo - the utxo to collect
   * @param typeScript - the type script to filter the cells
   */
  private async getRgbppCellsByUtxo(utxo: UTXO, typeScript?: Script): Promise<Cell[]> {
    try {
      const { txid, vout } = utxo;
      const args = buildRgbppLockArgs(vout, txid);

      const query: CKBIndexerQueryOptions = {
        lock: genRgbppLockScript(args, process.env.NETWORK === 'mainnet'),
      };

      if (typeScript) {
        query.type = typeScript;
      }

      const collector = this.cradle.ckb.indexer.collector(query).collect();
      const cells: Cell[] = [];
      for await (const cell of collector) {
        cells.push(cell);
      }
      return cells;
    } catch (e) {
      Sentry.withScope((scope) => {
        scope.captureException(e);
      });
      throw e;
    }
  }

  /**
   * Collect the cells for the utxos, return the utxo and the cells
   * @param utxos - the utxos to collect
   * @param typeScript - the type script to filter the cells
   */
  public async collectRgbppUtxoCellsPairs(utxos: UTXO[], typeScript?: Script): Promise<RgbppUtxoCellsPair[]> {
    const cells = await Promise.all(
      utxos.map((utxo) => {
        return this.limit(() =>
          asyncRetry(
            async () => {
              const cells = await this.getRgbppCellsByUtxo(utxo, typeScript);
              return { utxo, cells };
            },
            { retries: 2 },
          ),
        );
      }),
    );

    return cells.filter(({ cells }) => cells.length > 0);
  }

  /**
   * Enqueue a collect job to the queue
   * @param utxos - the utxos to collect
   */
  public async enqueueCollectJob(
    btcAddress: string,
    utxos: UTXO[],
    allowDuplicate?: boolean,
  ): Promise<Job<IRgbppCollectRequest>> {
    let jobId = btcAddress;
    if (allowDuplicate) {
      jobId = `${btcAddress}:${Date.now()}`;
    }
    return this.addJob(jobId, { btcAddress, utxos });
  }

  /**
   * Process the collect job, collect the cells for the utxos
   * concurrently controlled by the CKB_RPC_MAX_CONCURRENCY
   * retry 2 times if failed, and return the utxo and cells
   */
  public async process(job: Job<IRgbppCollectRequest>) {
    const { btcAddress, utxos } = job.data;
    const pairs = await this.collectRgbppUtxoCellsPairs(utxos);
    const data = pairs.reduce((acc, { utxo, cells }) => {
      const key = `${utxo.txid}:${utxo.vout}`;
      acc[key] = cells;
      return acc;
    }, {} as IRgbppCollectJobReturn);
    this.dataCahe.set(btcAddress, data);
    return data;
  }
}
