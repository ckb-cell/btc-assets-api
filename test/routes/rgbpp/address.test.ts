import { beforeEach, expect, test, vi } from 'vitest';
import { buildFastify } from '../../../src/app';
import { describe } from 'node:test';
import mockUTXOs from '../../__fixtures__/utxo.mock.json';
import mockRgbppUtxoPairs from '../../__fixtures__/rgbpp-utxo-pairs.mock.json';
import { RgbppUtxoCellsPair } from '../../../src/services/rgbpp';

let token: string;

describe('/rgbpp/v1/address', () => {
  beforeEach(async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/token/generate',
      payload: {
        app: 'test',
        domain: 'test.com',
      },
    });
    const data = response.json();
    token = data.token;

    await fastify.close();
  });

  test('/:btc_address/assets', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/assets',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('/:btc_address/balance - without pending_amount', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('/:btc_address/balance - with pending_amount', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );
    const transactionProcessor = fastify.container.resolve('transactionProcessor');
    const getPendingOuputCellsByTxidSpy = vi
      .spyOn(transactionProcessor, 'getPendingOuputCellsByTxid')
      .mockResolvedValueOnce([
        {
          cellOutput: {
            capacity: '0x5e9f52687',
            lock: {
              codeHash: '0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248',
              args: '0x010000000000000000000000000000000000000000000000000000000000000000000000',
              hashType: 'type',
            },
            type: {
              codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
              args: '0x30d3fbec9ceba691770d57c6d06bdb98cf0f82bef0ca6e87687a118d6ce1e7b7',
              hashType: 'type',
            },
          },
          data: '0x00e1f505000000000000000000000000',
        },
        {
          cellOutput: {
            capacity: '0x5e9f52687',
            lock: {
              codeHash: '0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248',
              args: '0x010000000000000000000000000000000000000000000000000000000000000000000000',
              hashType: 'type',
            },
            type: {
              codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
              args: '0x8c556e92974a8dd8237719020a259d606359ac2cc958cb8bda77a1c3bb3cd93b',
              hashType: 'type',
            },
          },
          data: '0x00e1f505000000000000000000000000',
        },
        {
          cellOutput: {
            capacity: '0x5e9f52687',
            lock: {
              codeHash: '0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248',
              args: '0x010000000000000000000000000000000000000000000000000000000000000001234567',
              hashType: 'type',
            },
            type: {
              codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
              args: '0x8c556e92974a8dd8237719020a259d606359ac2cc958cb8bda77a1c3bb3cd93b',
              hashType: 'type',
            },
          },
          data: '0x00e1f505000000000000000000000000',
        },
      ]);

    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(getPendingOuputCellsByTxidSpy).toBeCalledTimes(2);
    expect(getPendingOuputCellsByTxidSpy).toHaveBeenCalledWith(
      'aab2d8fc3f064087450057ccb6012893cf219043d8c915fe64c5322c0eeb6fd2',
    );
    expect(getPendingOuputCellsByTxidSpy).toHaveBeenCalledWith(
      '989f4e03179e17cbb6edd446f57ea6107a40ba23441056653f1cc34b7dd1e5ba',
    );
    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });
});
