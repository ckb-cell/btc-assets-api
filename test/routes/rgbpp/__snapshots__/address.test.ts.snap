// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/:btc_address/assets 1`] = `
[
  {
    "blockNumber": "0xc35b01",
    "cellOutput": {
      "capacity": "0x5e9f5203e",
      "lock": {
        "args": "0x0200000029ee57d8d3a4506b2647cea882a497da55f49dd4d71b41b14e28c651baa2050c",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x1ba116c119d1cfd98a53e9d1a615cf2af2bb87d95515c9d217d367054cfc696b",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0xc0d45407000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x4f8f6042b3646e677c871be1629150407b520e48174612d428506995f6055558",
    },
    "txIndex": "0x8",
  },
  {
    "blockNumber": "0xc87015",
    "cellOutput": {
      "capacity": "0x5e9e620d3",
      "lock": {
        "args": "0x02000000f42636cfd158d159c959ff931351636adff46b3cef5cd28d953b0fd07526d20b",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x1ba116c119d1cfd98a53e9d1a615cf2af2bb87d95515c9d217d367054cfc696b",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x008c8647000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xe098e5919e496fc6efb601307578cc436df7d554b9a3994f03b85d0ad2e8e7a6",
    },
    "txIndex": "0x1",
  },
  {
    "blockNumber": "0xcc3842",
    "cellOutput": {
      "capacity": "0xa7a358200",
      "lock": {
        "args": "0x000000008c8afb798a1637a284f1ef889e053a7f1e7030747f51cb86a31138c28b681aed",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xaf68a7bcf1423b57e7ce8636c28dedb463c6e69f62b1b35ca1ed9abf9a6b3d18",
        "codeHash": "0x685a60219309029d01310311dba953d67029170ca4848a4ff638e57002130a0d",
        "hashType": "data1",
      },
    },
    "data": "0xd40000001000000024000000b0000000100000006170706c69636174696f6e2f6a736f6e880000007b226e616d65223a22556e69636f726e20426f78222c227265736f75726365223a7b2275726c223a2268747470733a2f2f6172736565642e77656233696e6672612e6465762f306b4e437450376169417253596f6c6e424f656466705545493948554b7273323142443772495247735677222c2274797065223a22696d6167652f6a706567227d7d20000000ed3d643329453a418a8902a40c225ca81f405904b4ec9509ff09048240d30a86",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xdd0226408c4c31b2466f7c3d661d4e3e3c8b429631212a525f6668765baadc80",
    },
    "txIndex": "0x1",
  },
  {
    "blockNumber": "0xcb4375",
    "cellOutput": {
      "capacity": "0x5e9f53e00",
      "lock": {
        "args": "0x000000009b18c0f40491710e536b9916cb9d8326678b27ad5856cf51f0fdf031dec0f53e",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xc625c4ac0ba3ece5886d04958c3fc2d5558a21841c03577fad2d7c46e5b2b2b9",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x80969800000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x3aeb2a7f8e00ca32fae9b658da88274f2dcfeff41108ea28b567dc9b79c7dd3c",
    },
    "txIndex": "0x3",
  },
  {
    "blockNumber": "0xc35ba6",
    "cellOutput": {
      "capacity": "0x5e9f52f1f",
      "lock": {
        "args": "0x02000000bbd4d25df0dcc16d9e8cdd70872c1ce2a5acc4394a3cc32ab862ade346317d05",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x1ba116c119d1cfd98a53e9d1a615cf2af2bb87d95515c9d217d367054cfc696b",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0xa08d7313000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x71b4f66ac7f657e3b96b5b0694de5fadb61740d2644eabb91090dd9bce93316a",
    },
    "txIndex": "0x4",
  },
  {
    "blockNumber": "0xcb4378",
    "cellOutput": {
      "capacity": "0x5e9f53e00",
      "lock": {
        "args": "0x00000000b24761ad734952f78b5974e810f08a408965f27844772a35bfd6162dfe3eb4f7",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xc625c4ac0ba3ece5886d04958c3fc2d5558a21841c03577fad2d7c46e5b2b2b9",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x80f0fa02000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xb5ab02c1583387fa7bc3cb7e61daccead6d4d6b9a1294555328d4bc209ed5f8f",
    },
    "txIndex": "0x1",
  },
  {
    "blockNumber": "0xcc383d",
    "cellOutput": {
      "capacity": "0x94a39af00",
      "lock": {
        "args": "0x00000000906f620ad43ea196a10c077e9238515111cb0b6faa7eb69679e72528a17afafd",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xacb57628e8814a675ac3dfcdcd86b89da85399a82bfceedccb2c8a2200c48c1e",
        "codeHash": "0x685a60219309029d01310311dba953d67029170ca4848a4ff638e57002130a0d",
        "hashType": "data1",
      },
    },
    "data": "0xa100000010000000190000007d00000005000000646f622f30600000007b22626c6f636b5f6e756d626572223a31333038343932382c2263656c6c5f6964223a363132303533383831383036373235383030302c22646e61223a226166323566336565383563383734623030343638316334333162373835626231227d20000000c1097c85aa781d2cd3c017dc911eceab94644cfe2dd714ae7f7bbe4b75accb08",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x3b1a2d2d00a5ef95b45d93cc1d9394a411319f956121d95104b784e8b469f538",
    },
    "txIndex": "0x3",
  },
  {
    "blockNumber": "0xc35ba6",
    "cellOutput": {
      "capacity": "0x5e9f52f1f",
      "lock": {
        "args": "0x02000000976e945e25e9645ab5705c0bebe687e75f8b678fc6e0c9b26081b368f51563a4",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x1ba116c119d1cfd98a53e9d1a615cf2af2bb87d95515c9d217d367054cfc696b",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x01000000000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x274865dc9e754ab8fc96cbc7be20093feca0073b91d3157aed2639f6e89ada90",
    },
    "txIndex": "0x8",
  },
  {
    "blockNumber": "0xcc3d9d",
    "cellOutput": {
      "capacity": "0x5e9f53e00",
      "lock": {
        "args": "0x01000000a7fd61a66986273e1e8500e276e5ade9e1d66658e73cc1a4123a440ef6f78d36",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x30d3fbec9ceba691770d57c6d06bdb98cf0f82bef0ca6e87687a118d6ce1e7b7",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x00e1f505000000000000000000000000",
    "outPoint": {
      "index": "0x1",
      "txHash": "0xefcada73aa020bfc4c4dc7192f9a6a6e6152af77d1bb2c2380314b4130751bc4",
    },
    "txIndex": "0x6",
  },
  {
    "blockNumber": "0xcb42df",
    "cellOutput": {
      "capacity": "0x5e9f53e00",
      "lock": {
        "args": "0x00000000fe7e313c6bb99211e222dd1a4ae62a046ef4170d1e74a7252a827b0636ef6965",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xc625c4ac0ba3ece5886d04958c3fc2d5558a21841c03577fad2d7c46e5b2b2b9",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x00e1f505000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xeec5c357338f622f9bf586532be687f8fc1918566f27b4abb7fa5f9c1891af6b",
    },
    "txIndex": "0x2",
  },
  {
    "blockNumber": "0xcb669c",
    "cellOutput": {
      "capacity": "0x5e9f53e00",
      "lock": {
        "args": "0x000000004db3c41e5993ba73c0c916494d3f0e09042f2dcd8d48abd64b13a2ed6a149ffc",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0xc625c4ac0ba3ece5886d04958c3fc2d5558a21841c03577fad2d7c46e5b2b2b9",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0xa0860100000000000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0x2f6b86c4311dc854e779c0bbce110f397a0acc03dfe435aedf832b4f3de1a485",
    },
    "txIndex": "0x1",
  },
  {
    "blockNumber": "0xcb7072",
    "cellOutput": {
      "capacity": "0xa7a358200",
      "lock": {
        "args": "0x00000000ab39226c7b575b968798ebdb7f72b2ecec3d7cd6d71c38d11f095b8e2a8e61f3",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x76fd065c31d24a2c8219e7d56942fed3d49ff728697d98a1079edb06187ec91c",
        "codeHash": "0x685a60219309029d01310311dba953d67029170ca4848a4ff638e57002130a0d",
        "hashType": "data1",
      },
    },
    "data": "0xd40000001000000024000000b0000000100000006170706c69636174696f6e2f6a736f6e880000007b226e616d65223a22556e69636f726e20426f78222c227265736f75726365223a7b2275726c223a2268747470733a2f2f6172736565642e77656233696e6672612e6465762f306b4e437450376169417253596f6c6e424f656466705545493948554b7273323142443772495247735677222c2274797065223a22696d6167652f6a706567227d7d20000000ed3d643329453a418a8902a40c225ca81f405904b4ec9509ff09048240d30a86",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xbd1886d03da82d0a012af423044d8f192598d3a7be454397497fd22aba2a784e",
    },
    "txIndex": "0x1",
  },
]
`;

exports[`/:btc_address/balance - with pending_amount 1`] = `
{
  "address": "tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu",
  "xudt": [
    {
      "available_amount": "0x624eee61",
      "decimal": 8,
      "name": "",
      "pending_amount": "0x0",
      "symbol": "LP-UTXOSwap V1",
      "total_amount": "0x624eee61",
      "typeHash": "0x7337d2075479d5c6789564516be1db11478e1d3ca70ad88050fe0bf3ff05acc2",
    },
    {
      "available_amount": "0x98aeea0",
      "decimal": 8,
      "name": "JoyID Dex Test Inscription",
      "pending_amount": "0x0",
      "symbol": "JDI",
      "total_amount": "0x98aeea0",
      "typeHash": "0x6d7dd4497d1ef095644e422d30b2456cfe32c6469346c77752a4d6b56f432a73",
    },
    {
      "available_amount": "0x5f5e100",
      "decimal": 8,
      "name": "",
      "pending_amount": "0x5f5e100",
      "symbol": "LP-UTXOSwap V1",
      "total_amount": "0xbebc200",
      "typeHash": "0x78e21efcf107e7886eadeadecd1a01cfb88f1e5617f4438685db55b3a540d202",
    },
    {
      "available_amount": "0x0",
      "decimal": 8,
      "name": "XUDT Test Token",
      "pending_amount": "0x5f5e100",
      "symbol": "XTT",
      "total_amount": "0x5f5e100",
      "typeHash": "0x10f511f2efb0027191b97ac5b4bd77374ffdac7399e8527d76f5f9bd32e7d35b",
    },
  ],
}
`;

exports[`/:btc_address/balance - without pending_amount 1`] = `
{
  "address": "tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu",
  "xudt": [
    {
      "available_amount": "0x624eee61",
      "decimal": 8,
      "name": "",
      "pending_amount": "0x0",
      "symbol": "LP-UTXOSwap V1",
      "total_amount": "0x624eee61",
      "typeHash": "0x7337d2075479d5c6789564516be1db11478e1d3ca70ad88050fe0bf3ff05acc2",
    },
    {
      "available_amount": "0x98aeea0",
      "decimal": 8,
      "name": "JoyID Dex Test Inscription",
      "pending_amount": "0x0",
      "symbol": "JDI",
      "total_amount": "0x98aeea0",
      "typeHash": "0x6d7dd4497d1ef095644e422d30b2456cfe32c6469346c77752a4d6b56f432a73",
    },
    {
      "available_amount": "0x5f5e100",
      "decimal": 8,
      "name": "",
      "pending_amount": "0x0",
      "symbol": "LP-UTXOSwap V1",
      "total_amount": "0x5f5e100",
      "typeHash": "0x78e21efcf107e7886eadeadecd1a01cfb88f1e5617f4438685db55b3a540d202",
    },
  ],
}
`;
